### PR TITLE
Use narrowly scoped interfaces for client access

### PR DESCRIPTION
**Pull request by ironcladlou**
Use custom narrowly scoped interfaces for client access from the
RollingUpdater and Resizer. This allows for more flexible downstream
integration and unit testing without imposing a burden to implement
the entire client.Interface for just a handful of methods.

### CODE REVIEW

File: pkg/kubectl/resize.go [hunk 1]
```diff
@@ -89,7 +89,7 @@ type Resizer interface {
 	ResizeSimple(namespace, name string, preconditions *ResizePrecondition, newSize uint) (string, error)
 }
 
-func ResizerFor(kind string, c client.Interface) (Resizer, error) {
+func ResizerFor(kind string, c ResizerClient) (Resizer, error) {
```
Comment (thockin) [inline pkg/kubectl/resize.go]: FTR I think this is a pattern we should apply more.

File: pkg/kubectl/resize.go [hunk 2]
```diff
@@ -159,9 +158,33 @@ func (resizer *ReplicationControllerResizer) Resize(namespace, name string, newS
...
+
+// ResizerClient abstracts access to ReplicationControllers.
+type ResizerClient interface {
+	GetReplicationController(namespace, name string) (*api.ReplicationController, error)
+	UpdateReplicationController(namespace string, rc *api.ReplicationController) (*api.ReplicationController, error)
+	ControllerHasDesiredReplicas(rc *api.ReplicationController) wait.ConditionFunc
+}
+
+// RealResizerClient is a ResizerClient which uses a Kube client.
+type RealResizerClient struct {
```
Comment (thockin) [inline pkg/kubectl/resize.go]: I sort of hate that this is a public symbol.  Could we maybe define it in an internal pkg (using Go's internal trickery) to make shared across kubectl but not the whole universe?

Comment (pmorie) [inline pkg/kubectl/resize.go]: @thockin What internal trickery are you referring to?

File: pkg/kubectl/resize.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/resize.go]: Agreed, let's drop the hammer on anyone who even looks at *client.Client funny

File: pkg/kubectl/resize.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/kubectl/resize.go]: I think this is a good place to use a constructor method - leave the type private, but expose NewResizerClient(client.Interface)

Comment (ironcladlou) [inline pkg/kubectl/resize.go]: I like the constructor method. Let's see how that looks.

**Comment by ironcladlou** [comment]
@thockin @pmorie @smarterclayton 

https://github.com/GoogleCloudPlatform/kubernetes/commit/bd7b719944ca8c9f5d2f54b51b760f49a81abc91 introduces some constructors which allow us to keep the `real*` types package private. A definite improvement, IMO.

File: pkg/kubectl/resize.go (hunk 1 above)
Comment (pmorie) [inline pkg/kubectl/resize.go]: :hammer:

File: pkg/kubectl/resize.go (hunk 2 above)
Comment (thockin) [inline pkg/kubectl/resize.go]: FTR, Go trickery is not really available yet, so mea culpa:

https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit

**Comment by thockin** [comment]
I see how it works.  You drive-by some comments and BAM, you get assigned the code review.  That'll teach me.

LGTM

**Comment by smarterclayton** [comment]
You must be new here.

**Pull request merged**